File: 0201-0300/0264-Ugly Number II/0264-Ugly Number II.py
class Solution:
    def nthUglyNumber(self, n: int) -> int:
        p2 = p3 = p5 = 0
        ugly = [1] * n
        for i in range(1, n):
            ugly[i] = min(ugly[p2]*2, ugly[p3]*3, ugly[p5]*5)
            if ugly[p2]*2 == ugly[i]:
                p2 += 1
            if ugly[p3]*3 == ugly[i]:
                p3 += 1
            if ugly[p5]*5 == ugly[i]:
                p5 += 1
        return ugly[-1]

File: 0201-0300/0264-Ugly Number II/0264-Ugly Number II.cpp
class Solution {
public:
    int nthUglyNumber(int n) {
        int p2 = 0, p3 = 0, p5 = 0;
        vector<int> ugly(n, 1);
        for (int i = 1; i < n; ++i) {
            ugly[i] = min({ugly[p2] * 2, ugly[p3] * 3, ugly[p5] * 5});
            if (ugly[p2] * 2 == ugly[i]) {
                ++p2;
            }
            
            if (ugly[p3] * 3 == ugly[i]) {
                ++p3;
            }
            
            if (ugly[p5] * 5 == ugly[i]) {
                ++p5;
            }
        }
        
        return ugly[n - 1];        
    }
};
